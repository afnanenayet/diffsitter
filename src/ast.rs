//! Utilities for processing the ASTs provided by `tree_sitter`

use crate::diff::DiffEngine;
use crate::diff::Hunks;
use crate::diff::Myers;
use logging_timer::time;
use std::{cell::RefCell, ops::Index, path::PathBuf};
use tree_sitter::Node as TSNode;
use tree_sitter::Tree as TSTree;

/// A mapping between a tree-sitter node and the text it corresponds to
#[derive(Debug, Clone, Copy)]
pub struct Entry<'a> {
    /// The node an entry in the diff vector refers to
    ///
    /// We keep a reference to the leaf node so that we can easily grab the text and other metadata
    /// surrounding the syntax
    pub reference: TSNode<'a>,

    /// A reference to the text the node refers to
    ///
    /// This is different from the `source_text` that the [AstVector](AstVector) refers to, as the
    /// entry only holds a reference to the specific range of text that the node covers.
    pub text: &'a str,
}

/// A vector that allows for linear traversal through the leafs of an AST.
///
/// This representation of the tree leaves is much more convenient for things like dynamic
/// programming, and provides useful for formatting.
#[derive(Debug)]
pub struct AstVector<'a> {
    /// The leaves of the AST, build with an in-order traversal
    pub leaves: Vec<Entry<'a>>,

    /// The full source text that the AST refers to
    pub source_text: &'a str,
}

impl<'a> Eq for Entry<'a> {}

/// A wrapper struct for AST vector data that owns the data that the AST vector references
///
/// Ideally we would just have the AST vector own the actual string and tree, but it makes things
/// extremely messy with the borrow checker, so we have this wrapper struct that holds the owned
/// data that the vector references. This gets tricky because the tree sitter library uses FFI so
/// the lifetime references get even more mangled.
#[derive(Debug)]
pub struct AstVectorData {
    /// The text in the file
    pub text: String,

    /// The tree that was parsed using the text
    pub tree: TSTree,

    /// The file path that the text corresponds to
    pub path: PathBuf,
}

impl<'a> AstVector<'a> {
    /// Create a `DiffVector` from a `tree_sitter` tree
    ///
    /// This method calls a helper function that does an in-order traversal of the tree and adds
    /// leaf nodes to a vector
    #[time("info", "ast::{}")]
    pub fn from_ts_tree(tree: &'a TSTree, text: &'a str) -> Self {
        let leaves = RefCell::new(Vec::new());
        build(&leaves, tree.root_node(), text);
        AstVector {
            leaves: leaves.into_inner(),
            source_text: text,
        }
    }

    /// Return the number of nodes in the diff vector
    pub fn len(&self) -> usize {
        self.leaves.len()
    }
}

impl<'a> Index<usize> for AstVector<'a> {
    type Output = Entry<'a>;

    fn index(&self, index: usize) -> &Self::Output {
        &self.leaves[index]
    }
}

impl<'a> PartialEq for Entry<'a> {
    fn eq(&self, other: &Entry) -> bool {
        self.text == other.text
    }
}

impl<'a> PartialEq for AstVector<'a> {
    fn eq(&self, other: &AstVector) -> bool {
        if self.leaves.len() != other.leaves.len() {
            return false;
        }

        // Zip through each entry to determine whether the elements are equal. We start with a
        // `false` value for not equal and accumulate any inequalities along the way.
        let not_equal = self
            .leaves
            .iter()
            .zip(other.leaves.iter())
            .fold(false, |not_equal, (entry_a, entry_b)| {
                not_equal | (entry_a != entry_b)
            });
        !not_equal
    }
}

/// Recursively build a vector from a given node
///
/// This is a helper function that simply walks the tree and collects leaves in an in-order manner.
/// Every time it encounters a leaf node, it stores the metadata and reference to the node in an
/// `Entry` struct.
fn build<'a>(vector: &RefCell<Vec<Entry<'a>>>, node: tree_sitter::Node<'a>, text: &'a str) {
    // If the node is a leaf, we can stop traversing
    if node.child_count() == 0 {
        // We only push an entry if the referenced text range isn't empty, since there's no point
        // in having an empty text range. This also fixes a bug where the program would panic
        // because it would attempt to access the 0th index in an empty text range.
        if !node.byte_range().is_empty() {
            let node_text: &'a str = &text[node.byte_range()];
            vector.borrow_mut().push(Entry {
                reference: node,
                text: node_text,
            });
        }
        return;
    }

    let mut cursor = node.walk();

    for child in node.children(&mut cursor) {
        build(vector, child, text);
    }
}

/// The different types of elements that can be in an edit script
#[derive(Debug, Eq, PartialEq)]
pub enum EditType<T> {
    /// An element that was added in the edit script
    Addition(T),

    /// An element that was deleted in the edit script
    Deletion(T),
}

/// Compute the hunks corresponding to the minimum edit path between two documents
///
/// This method computes the minimum edit distance between two `DiffVector`s, which are the leaf
/// nodes of an AST, using the standard DP approach to the longest common subsequence problem, the
/// only twist is that here, instead of operating on raw text, we're operating on the leaves of an
/// AST.
///
/// This has O(mn) space complexity and uses O(mn) space to compute the minimum edit path, and then
/// has O(mn) space complexity and uses O(mn) space to backtrack and recreate the path.
///
/// This will return two groups of [hunks](diff::Hunks) in a tuple of the form
/// `(old_hunks, new_hunks)`.
#[time("info", "ast::{}")]
pub fn compute_edit_script<'a>(a: &'a AstVector, b: &'a AstVector) -> (Hunks<'a>, Hunks<'a>) {
    let myers = Myers::default();
    let edit_script = myers.diff(&a.leaves[..], &b.leaves[..]);
    let mut old_edits = Vec::with_capacity(edit_script.len());
    let mut new_edits = Vec::with_capacity(edit_script.len());

    for edit in edit_script {
        match edit {
            EditType::Deletion(&edit) => old_edits.push(edit),
            EditType::Addition(&edit) => new_edits.push(edit),
        }
    }

    // Convert the vectors of edits into hunks that can be displayed
    let old_hunks = old_edits.into_iter().collect();
    let new_hunks = new_edits.into_iter().collect();
    (old_hunks, new_hunks)
}
