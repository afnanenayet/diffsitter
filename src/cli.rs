//! Code related to the CLI

use crate::parse::supported_languages;
use std::path::PathBuf;
use structopt::StructOpt;
use strum_macros::{EnumString, ToString};

#[cfg(target_os = "macos")]
use std::env;

#[cfg(not(target_os = "macos"))]
use directories;

/// The environment variable for XDG_CONFIG
#[cfg(target_os = "macos")]
const XDG_CONFIG: &str = "XDG_CONFIG";
/// The directory inside the config base where the config file is stored
const CFG_DIRECTORY: &str = "diffsitter";
/// The expected filename for the config file
const CFG_FILE_NAME: &str = "config.toml";

#[derive(Debug, Eq, PartialEq, Clone, StructOpt)]
#[structopt(
    name = "diffsitter",
    about = "AST based diffs",
    setting = structopt::clap::AppSettings::ColoredHelp
)]
pub struct Args {
    /// Print debug output
    ///
    /// This will print debug output
    #[structopt(short, long)]
    pub debug: bool,
    /// Run a subcommand that doesn't perform a diff. Valid options are: "list",
    /// "dump_default_config".
    ///
    /// "list" will list all of the filetypes/languages that this program was compiled with support
    /// for. "dump_default_config" will dump the default configuration to stdout.
    #[structopt(long)]
    pub cmd: Option<Command>,
    /// The first file to compare against
    ///
    /// Text that is in this file but is not in the new file is considered a deletion
    #[structopt(name = "OLD", parse(from_os_str), required_unless = "cmd")]
    pub old: Option<PathBuf>,
    /// The file that the old file is compared against
    ///
    /// Text that is in this file but is not in the old file is considered an addition
    #[structopt(name = "NEW", parse(from_os_str), required_unless = "cmd")]
    pub new: Option<PathBuf>,
    /// Manually set the file type for the given files
    ///
    /// This will dictate which parser is used with the difftool. You can list all of the valid
    /// file type strings with `diffsitter --cmd list`
    #[structopt(short = "t", long = "filetype")]
    pub file_type: Option<String>,
    /// Use the config provided at the given path
    ///
    /// By default, diffsitter attempts to find the config at $XDG_CONFIG/diffsitter.toml
    #[structopt(short, long, env = "DIFFSITTER_CONFIG")]
    pub config: Option<PathBuf>,
    /// Set the color output policy. Valid values are: "auto", "on", "off".
    ///
    /// "auto" will automatically detect whether colors should be applied by trying to determine
    /// whether the process is outputting to a TTY. "on" will force color output. "off" will
    /// forcibly disable color output.
    #[structopt(long = "color", default_value)]
    pub color_output: ColorOutputPolicy,
    /// Don't read any configs and use the default settings
    ///
    /// This will cause the app to ignore any configs and all values will pull from the default
    /// settings. This cannot be
    #[structopt(short, long)]
    pub no_config: bool,
}

/// Commands related to the configuration
#[derive(Debug, Eq, PartialEq, Clone, Copy, StructOpt, EnumString)]
#[strum(serialize_all = "snake_case")]
pub enum Command {
    /// List the languages that this program was compiled for
    List,
    /// Dump the default config to stdout
    DumpDefaultConfig,
}

/// Print a list of the languages that this instance of diffsitter was compiled with
pub fn list_supported_languages() {
    let languages = supported_languages();

    println!("This program was compiled with support for:");

    for language in languages {
        println!("- {}", language);
    }
}

/// Return the default location for the config file (for *nix, Linux and MacOS), this will use
/// $XDG_CONFIG/.config, where `$XDG_CONFIG` is `$HOME/.config` by default.
#[cfg(target_os = "macos")]
pub fn default_config_file() -> PathBuf {
    // First we check to see if $XDG_CONfIG is set
    let mut config_dir = if let Ok(dir) = env::var(XDG_CONFIG) {
        PathBuf::from(dir)
    } else {
        let base_dirs = directories::BaseDirs::new().unwrap();
        let home_dir = base_dirs.home_dir();
        let mut path = PathBuf::from(home_dir);
        path.push(".config");
        path
    };
    config_dir.push(CFG_DIRECTORY);
    config_dir.push(CFG_FILE_NAME);
    config_dir
}

/// Return the default location for the config file (for windows), this will use
/// $XDG_CONFIG/.config, where `$XDG_CONFIG` is `$HOME/.config` by default.
#[cfg(not(target_os = "macos"))]
pub fn default_config_file() -> PathBuf {
    let base_dirs = directories::BaseDirs::new().unwrap();
    let mut config_file: PathBuf = base_dirs.config_dir().into();
    config_file.push(CFG_DIRECTORY);
    config_file.push(CFG_FILE_NAME);
    config_file
}

/// Whether the output to the terminal should be colored
#[derive(Debug, Clone, Copy, PartialEq, Eq, EnumString, ToString)]
#[strum(serialize_all = "snake_case")]
pub enum ColorOutputPolicy {
    /// Automatically enable color if printing to a TTY, otherwise disable color
    Auto,
    /// Force plaintext output
    Off,
    /// Force color output
    On,
}

impl Default for ColorOutputPolicy {
    fn default() -> Self {
        ColorOutputPolicy::Auto
    }
}

/// Set whether the terminal should display colors based on the user's preferences
///
/// This method will set the terminal output policy *for the current thread*.
pub fn set_term_colors(color_opt: ColorOutputPolicy) {
    match color_opt {
        ColorOutputPolicy::Off => (console::set_colors_enabled(false)),
        ColorOutputPolicy::On => (console::set_colors_enabled(true)),
        _ => (),
    };
}
