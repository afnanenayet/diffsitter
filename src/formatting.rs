//! Utilities related to displaying/formatting the edits computed as the difference between two
//! ASTs

use crate::diff::{Hunk, Hunks, Line};
use anyhow::Result;
use console::{Color, Style, Term};
use log::debug;
use serde::{Deserialize, Serialize};
use std::{cmp::Ordering, io::Write};
use strum_macros::EnumString;

/// A copy of the [Color](console::Color) enum so we can serialize using serde, and get around the
/// orphan rule.
#[derive(Debug, PartialEq, Eq, Copy, Clone, Serialize, Deserialize)]
#[serde(remote = "Color", rename_all = "snake_case")]
enum ColorDef {
    Color256(u8),
    Black,
    Red,
    Green,
    Yellow,
    Blue,
    Magenta,
    Cyan,
    White,
}

impl From<ColorDef> for Color {
    fn from(c: ColorDef) -> Self {
        match c {
            ColorDef::Black => Color::Black,
            ColorDef::White => Color::White,
            ColorDef::Red => Color::Red,
            ColorDef::Green => Color::Green,
            ColorDef::Yellow => Color::Yellow,
            ColorDef::Blue => Color::Blue,
            ColorDef::Magenta => Color::Magenta,
            ColorDef::Cyan => Color::White,
            ColorDef::Color256(c) => Color::Color256(c),
        }
    }
}

impl Default for ColorDef {
    fn default() -> Self {
        ColorDef::Black
    }
}

/// Formatting directives for text
///
/// This was abstracted out because the exact same settings apply for both additions and deletions
#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct TextFormatting {
    /// The highlight/background color to use with emphasized text
    #[serde(with = "opt_color_def", default = "default_option")]
    pub highlight: Option<Color>,
    /// The foreground color to use with un-emphasized text
    #[serde(with = "ColorDef")]
    pub regular_foreground: Color,
    /// The foreground color to use with emphasized text
    #[serde(with = "ColorDef")]
    pub emphasized_foreground: Color,
    /// Whether to bold emphasized text
    pub bold: bool,
    /// Whether to underline emphasized text
    pub underline: bool,
    /// The prefix to use with the line
    pub prefix: String,
}

/// A helper function for the serde serializer
///
/// Due to the shenanigans we're using to serialize the optional color, we need to supply this
/// method so serde can infer a default value for an option when its key is missing.
fn default_option<T>() -> Option<T> {
    None
}

impl TextFormatting {
    /// Generate a [Style] for regular text
    pub fn to_regular_style(&self) -> Style {
        let mut style = Style::default();
        style = style.fg(self.regular_foreground);
        style
    }

    /// Generate a [Style] for emphasized text
    pub fn to_emphasized_style(&self) -> Style {
        let mut style = Style::default();
        style = style.fg(self.emphasized_foreground);

        if self.bold {
            style = style.bold();
        }

        if self.underline {
            style = style.underlined();
        }

        if let Some(color) = self.highlight {
            style = style.bg(color);
        }
        style
    }
}

/// Formatting options for rendering a diff
#[derive(Debug, PartialEq, Serialize, Deserialize, Clone)]
#[serde(default, rename_all = "kebab-case")]
pub struct Options {
    /// The formatting options to use with text addition
    pub addition: TextFormatting,
    /// The formatting options to use with text addition
    pub deletion: TextFormatting,
}

impl Default for Options {
    fn default() -> Self {
        Options {
            addition: TextFormatting {
                regular_foreground: Color::Green,
                emphasized_foreground: Color::Green,
                highlight: None,
                bold: true,
                underline: false,
                prefix: "+ ".into(),
            },
            deletion: TextFormatting {
                regular_foreground: Color::Red,
                emphasized_foreground: Color::Red,
                highlight: None,
                bold: true,
                underline: false,
                prefix: "- ".into(),
            },
        }
    }
}

/// User supplied parameters that are required to display a diff
#[derive(Debug)]
pub struct DisplayParameters<'text> {
    /// The hunks corresponding to the old document
    pub old_hunks: &'text Hunks<'text>,
    /// The hunks corresponding to the old document
    pub new_hunks: &'text Hunks<'text>,
    /// The full text of the old document
    pub old_text: &'text str,
    /// The full text of the new document
    pub new_text: &'text str,
    /// The file name of the old document
    pub old_text_filename: &'text str,
    /// The file name of the new document
    pub new_text_filename: &'text str,
}

/// The formatting directives to use with different types of text
///
/// This struct defines the different types of text to use with `print_line`.
struct FormattingDirectives<'a> {
    /// The formatting to use with normal unchanged text in a diff line
    pub regular: &'a Style,
    /// The formatting to use with emphasized text in a diff line
    pub emphasis: &'a Style,
    /// The prefix (if any) to use with the line
    pub prefix: &'a str,
}

impl Options {
    /// A helper function for printing a line-by-line diff
    ///
    /// This will process the "raw" [diff vector](AstVector) and turn extract the differences
    /// between lines.
    // TODO(afnan) make this function private and use `print` to dispatch
    pub fn print(&self, term: &mut Term, params: &DisplayParameters) -> Result<()> {
        let &DisplayParameters {
            old_text,
            new_text,
            old_hunks,
            new_hunks,
            old_text_filename,
            new_text_filename,
        } = params;
        let old_fmt = FormattingDirectives {
            regular: &self.deletion.to_regular_style(),
            emphasis: &self.deletion.to_emphasized_style(),
            prefix: &self.deletion.prefix,
        };
        let new_fmt = FormattingDirectives {
            regular: &self.addition.to_regular_style(),
            emphasis: &self.addition.to_emphasized_style(),
            prefix: &self.addition.prefix,
        };

        // We need access to specific line numbers in the text so we can print out text ranges
        // within a line. It's more efficient to break up the text by line up-front so we don't
        // have to re-do that when we print out each line/hunk.
        let old_lines: Vec<_> = old_text.lines().collect();
        let new_lines: Vec<_> = new_text.lines().collect();

        self.print_title(
            term,
            old_text_filename,
            new_text_filename,
            &old_fmt,
            &new_fmt,
        )?;

        // Iterate through the edits on both documents. We know that both of the vectors are
        // sorted, and we can use that property to iterate through the entries in O(n). Basic
        // leetcode woo!
        let mut it_old = 0;
        let mut it_new = 0;

        while it_old < old_hunks.0.len() && it_new < new_hunks.0.len() {
            let old_hunk = &old_hunks.0[it_old];
            let new_hunk = &new_hunks.0[it_new];

            // We can unwrap here because the loop invariant enforces that there is at least one
            // element in the deque, otherwise the loop wouldn't run at all.
            let old_line_num = old_hunk.first_line().unwrap();
            let new_line_num = new_hunk.first_line().unwrap();

            match old_line_num.cmp(&new_line_num) {
                Ordering::Equal => {
                    self.print_hunk(term, &old_lines, old_hunk, &old_fmt)?;
                    self.print_hunk(term, &new_lines, new_hunk, &new_fmt)?;
                    it_old += 1;
                    it_new += 1;
                }
                Ordering::Less => {
                    self.print_hunk(term, &old_lines, old_hunk, &old_fmt)?;
                    it_old += 1;
                }
                Ordering::Greater => {
                    self.print_hunk(term, &new_lines, new_hunk, &new_fmt)?;
                    it_new += 1;
                }
            };
        }

        debug!("Printing remaining old hunks");

        while it_old < old_hunks.0.len() {
            let hunk = &old_hunks.0[it_old];
            self.print_hunk(term, &old_lines, hunk, &old_fmt)?;
            it_old += 1;
        }

        debug!("Printing remaining new hunks");

        while it_new < new_hunks.0.len() {
            let hunk = &new_hunks.0[it_new];
            self.print_hunk(term, &new_lines, hunk, &new_fmt)?;
            it_new += 1;
        }
        Ok(())
    }

    /// Print the title for the diff
    fn print_title(
        &self,
        term: &mut Term,
        old_fname: &str,
        new_fname: &str,
        old_fmt: &FormattingDirectives,
        new_fmt: &FormattingDirectives,
    ) -> std::io::Result<()> {
        let divider = " -> ";
        write!(
            term,
            "{}{}{}\n",
            old_fmt.regular.apply_to(old_fname),
            divider,
            new_fmt.regular.apply_to(new_fname)
        )?;
        // We get the sizes of the individual strings rather than just take the size of the string
        // that we pass into the write method above
        let sep_size = old_fname.len() + divider.len() + new_fname.len();
        write!(term, "{}\n", "-".repeat(sep_size))?;
        Ok(())
    }

    /// Print a [hunk](Hunk) to `stdout`
    fn print_hunk(
        &self,
        term: &mut Term,
        lines: &[&str],
        hunk: &Hunk,
        fmt: &FormattingDirectives,
    ) -> Result<()> {
        debug!(
            "Printing hunk (lines {} - {})",
            hunk.first_line().unwrap(),
            hunk.last_line().unwrap()
        );

        for line in &hunk.0 {
            let text = lines[line.line_index];
            debug!("Printing line {}", line.line_index);
            self.print_line(term, text, line, fmt)?;
            debug!("End line {}", line.line_index);
        }
        debug!(
            "End hunk (lines {} - {})",
            hunk.first_line().unwrap(),
            hunk.last_line().unwrap()
        );
        Ok(())
    }

    /// Print a line with edits
    ///
    /// This is a generic helper method for additions and deletions, since the logic is very
    /// similar, they just use different styles.
    ///
    /// `text` refers to the text that corresponds line number of the given [line](Line).
    fn print_line(
        &self,
        term: &mut Term,
        text: &str,
        line: &Line,
        fmt: &FormattingDirectives,
    ) -> Result<()> {
        // First, we print the prefix to stdout
        term.write_str(&fmt.regular.apply_to(fmt.prefix).to_string())?;

        // The number of characters that have been printed out to stdout already. These aren't
        // *actually* chars because UTF-8, but you get the gist.
        let mut printed_chars = 0;

        // We keep printing ranges until we've covered the entire line
        for entry in &line.entries {
            // The range of text to emphasize
            // TODO(afnan) deal with ranges spanning multiple rows
            let emphasis_range =
                entry.reference.start_position().column..entry.reference.end_position().column;

            // First we need to see if there's any regular text to cover. If the range has a len of
            // zero this is a no-op
            let regular_range = printed_chars..emphasis_range.start;
            let regular_text: String = text[regular_range].into();
            term.write_str(&fmt.regular.apply_to(regular_text).to_string())?;

            // Need to set the printed_chars marker here because emphasized_text moves the range
            printed_chars = emphasis_range.end;
            let emphasized_text: String = text[emphasis_range].into();
            term.write_str(&fmt.emphasis.apply_to(emphasized_text).to_string())?;
        }
        // Finally, print any normal text after the last entry
        let remaining_range = printed_chars..text.len();
        let remaining_text: String = text[remaining_range].into();
        term.write_str(&fmt.regular.apply_to(remaining_text).to_string())?;
        term.write_str("\n")?;
        Ok(())
    }
}

/// The method to use when emphasizing diffs within a line
///
/// `Bold` is used as the default emphasis strategy between two lines.
#[derive(Debug, PartialEq, EnumString, Serialize, Deserialize)]
#[strum(serialize_all = "snake_case")]
pub enum Emphasis {
    /// Don't emphasize anything
    ///
    /// This field exists because the absence of a value implies that the user wants to use the
    /// default emphasis strategy.
    None,
    /// Bold the differences between the two lines for emphasis
    Bold,
    /// Underline the differences between two lines for emphasis
    Underline,
    /// Use a colored highlight for emphasis
    Highlight(HighlightColors),
}

impl Default for Emphasis {
    fn default() -> Self {
        Emphasis::Bold
    }
}

/// Specify the colors to use when highlighting differences
// TODO(afnan) implement the proper defaults for this struct
#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct HighlightColors {
    /// The background color to use with an addition
    #[serde(with = "ColorDef")]
    pub addition: Color,
    /// The background color to use with a deletion
    #[serde(with = "ColorDef")]
    pub deletion: Color,
}

impl Default for HighlightColors {
    fn default() -> Self {
        HighlightColors {
            addition: Color::Color256(0),
            deletion: Color::Color256(0),
        }
    }
}

// Workaround so we can use the `ColorDef` remote serialization mechanism with optional types
mod opt_color_def {
    use super::{Color, ColorDef};
    use serde::{Deserialize, Deserializer, Serialize, Serializer};

    pub fn serialize<S>(value: &Option<Color>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Helper<'a>(#[serde(with = "ColorDef")] &'a Color);

        value.as_ref().map(Helper).serialize(serializer)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<Color>, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct Helper(#[serde(with = "ColorDef")] Color);

        let helper = Option::deserialize(deserializer)?;
        Ok(helper.map(|Helper(external)| external))
    }
}
