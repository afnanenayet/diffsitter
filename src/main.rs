mod ast;
mod cli;
mod diff;
mod formatting;
mod parse;

use anyhow::{anyhow, Result};
use ast::AstVector;
use cli::{default_config_file, list_supported_languages, set_term_colors, Args};
use console::Term;
use formatting::DisplayParameters;
use formatting::Options;
use log::{info, LevelFilter};
use std::fs;

/// Return an instance of [Options] froma config file path
///
/// If a config path isn't provided, then this will use the [default
/// path](cli::default_config_file).
fn derive_options(args: &Args) -> Result<Options> {
    if args.no_config {
        info!("`no_config` specified -- falling back to optional config");
        return Ok(Options::default());
    }
    let config_fp = if let Some(path) = args.config.as_ref() {
        path.clone()
    } else {
        default_config_file()
    };
    info!("Reading config at {:#?}", config_fp);

    if let Ok(config_str) = fs::read_to_string(config_fp) {
        return toml::from_str(&config_str).map_err(|e| anyhow!(e));
    }
    Ok(Options::default())
}

/// Take the diff of two files
fn run_diff(args: &Args) -> Result<()> {
    let options = derive_options(args)?;
    let path_a = args.old.as_ref().unwrap();
    let path_b = args.new.as_ref().unwrap();

    let old_text = fs::read_to_string(&path_a)?;
    info!("Reading {:#?} to string", &path_a);
    let new_text = fs::read_to_string(&path_b)?;
    info!("Reading {:#?} to string", &path_b);
    let file_type: Option<&str> = args.file_type.as_deref();

    if let Some(file_type) = file_type {
        info!("Using user-set filetype: {}", file_type);
    } else {
        info!("Will deduce filetype from file extension");
    }
    let ast_a = parse::parse_file(&path_a, file_type)?;
    let ast_b = parse::parse_file(&path_b, file_type)?;
    let diff_vec_a = AstVector::from_ts_tree(&ast_a, &old_text);
    let diff_vec_b = AstVector::from_ts_tree(&ast_b, &new_text);
    let (old_hunks, new_hunks) = ast::edit_hunks(&diff_vec_a, &diff_vec_b)?;
    let params = DisplayParameters {
        old_hunks: &old_hunks,
        new_hunks: &new_hunks,
        old_text: &old_text,
        new_text: &new_text,
    };
    let mut term = Term::stdout();
    options.print(&mut term, &params)?;
    Ok(())
}

/// Serialize the default options struct to a TOML file and print that to stdout
fn dump_default_config() -> Result<()> {
    let config = Options::default();
    let s = toml::to_string_pretty(&config)?;
    println!("{}", s);
    Ok(())
}

#[paw::main]
fn main(args: Args) -> Result<()> {
    use cli::Command;

    let log_level = if args.debug {
        LevelFilter::Trace
    } else {
        LevelFilter::Off
    };
    pretty_env_logger::formatted_timed_builder()
        .filter_level(log_level)
        .init();

    if let Some(cmd) = args.cmd {
        match cmd {
            Command::List => list_supported_languages(),
            Command::DumpDefaultConfig => dump_default_config()?,
        }
    } else {
        set_term_colors(args.color_output);
        run_diff(&args)?;
    }
    Ok(())
}
