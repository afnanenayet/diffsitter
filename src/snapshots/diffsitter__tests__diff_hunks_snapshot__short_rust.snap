---
source: src/main.rs
assertion_line: 305
expression: diff_hunks

---
(
    Hunks(
        [
            Hunk(
                [
                    Line {
                        line_index: 1,
                        entries: [
                            Entry {
                                reference: {Node let (1, 4) - (1, 7)},
                                text: "let",
                            },
                            Entry {
                                reference: {Node identifier (1, 8) - (1, 9)},
                                text: "x",
                            },
                            Entry {
                                reference: {Node = (1, 10) - (1, 11)},
                                text: "=",
                            },
                            Entry {
                                reference: {Node integer_literal (1, 12) - (1, 13)},
                                text: "1",
                            },
                            Entry {
                                reference: {Node ; (1, 13) - (1, 14)},
                                text: ";",
                            },
                        ],
                    },
                ],
            ),
            Hunk(
                [
                    Line {
                        line_index: 4,
                        entries: [
                            Entry {
                                reference: {Node identifier (4, 3) - (4, 10)},
                                text: "add_one",
                            },
                        ],
                    },
                ],
            ),
        ],
    ),
    Hunks(
        [
            Hunk(
                [
                    Line {
                        line_index: 9,
                        entries: [
                            Entry {
                                reference: {Node } (9, 0) - (9, 1)},
                                text: "}",
                            },
                        ],
                    },
                ],
            ),
            Hunk(
                [
                    Line {
                        line_index: 11,
                        entries: [
                            Entry {
                                reference: {Node fn (11, 0) - (11, 2)},
                                text: "fn",
                            },
                            Entry {
                                reference: {Node identifier (11, 3) - (11, 11)},
                                text: "addition",
                            },
                            Entry {
                                reference: {Node ( (11, 11) - (11, 12)},
                                text: "(",
                            },
                            Entry {
                                reference: {Node ) (11, 12) - (11, 13)},
                                text: ")",
                            },
                            Entry {
                                reference: {Node { (11, 14) - (11, 15)},
                                text: "{",
                            },
                        ],
                    },
                ],
            ),
            Hunk(
                [
                    Line {
                        line_index: 14,
                        entries: [
                            Entry {
                                reference: {Node identifier (14, 3) - (14, 10)},
                                text: "add_two",
                            },
                            Entry {
                                reference: {Node ( (14, 10) - (14, 11)},
                                text: "(",
                            },
                            Entry {
                                reference: {Node ) (14, 11) - (14, 12)},
                                text: ")",
                            },
                        ],
                    },
                ],
            ),
        ],
    ),
)
